pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	once                sync.Once
+	upgrader            ClusterUpgrader
+	UpgradeStepOrdering = []upgradev1alpha1.UpgradeConditionType{
+		upgradev1alpha1.UpgradeValidated,
+		upgradev1alpha1.UpgradePreHealthCheck,
+		upgradev1alpha1.UpgradeScaleUpExtraNodes,
+		upgradev1alpha1.CommenceUpgrade,
+		upgradev1alpha1.ControlPlaneMaintWindow,
+		upgradev1alpha1.ControlPlaneUpgraded,
+		upgradev1alpha1.AllMastersUpgraded,
+		upgradev1alpha1.RemoveControlPlaneMaintWindow,
+		upgradev1alpha1.WorkersMaintWindow,
+		upgradev1alpha1.AllWorkerUpgraded,
+		upgradev1alpha1.PostUpgradeFix,
+		upgradev1alpha1.RemoveExtraScaledNodes,
+		upgradev1alpha1.UpdateSubscriptions,
+		upgradev1alpha1.PostUpgradeVerification,
+		upgradev1alpha1.RemoveMaintWindow,
+		upgradev1alpha1.PostClusterHealthCheck,
+	}
+)
+
+const (
+	TIMEOUT_SCALE_EXTRAL_NODES = 30 * time.Minute
+
+	LABEL_UPGRADE = "upgrade.managed.openshift.io"
+)
+
+type ClusterUpgrader map[upgradev1alpha1.UpgradeConditionType]UpgradeStep
+
+// Ordering returns the ordering of predicates.
+func Ordering() []upgradev1alpha1.UpgradeConditionType {
+	return UpgradeStepOrdering
+}
+
+type UpgradeStep func(client client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error)
+
+func NewUpgrader(reqLogger logr.Logger, client client.Client) ClusterUpgrader {
+	once.Do(func() {
+		upgrader = map[upgradev1alpha1.UpgradeConditionType]UpgradeStep{
+			upgradev1alpha1.UpgradeValidated:              ValidateUpgradeConfig,
+			upgradev1alpha1.UpgradePreHealthCheck:         PreClusterHealthCheck,
+			upgradev1alpha1.UpgradeScaleUpExtraNodes:      EnsureExtraUpgradeWorkers,
+			upgradev1alpha1.ControlPlaneMaintWindow:       CreateControlPlaneMaintWindow,
+			upgradev1alpha1.CommenceUpgrade:               CommenceUpgrade,
+			upgradev1alpha1.ControlPlaneUpgraded:          ControlPlaneUpgraded,
+			upgradev1alpha1.AllMastersUpgraded:            AllMastersUpgraded,
+			upgradev1alpha1.RemoveControlPlaneMaintWindow: RemoveControlPlaneMaintWindow,
+			upgradev1alpha1.WorkersMaintWindow:            CreateWorkerMaintWindow,
+			upgradev1alpha1.AllWorkerUpgraded:             AllWorkersUpgraded,
+			upgradev1alpha1.PostUpgradeFix:                PostUpgradeFix,
+			upgradev1alpha1.RemoveExtraScaledNodes:        RemoveExtraScaledNodes,
+			upgradev1alpha1.UpdateSubscriptions:           UpdateSubscriptions,
+			upgradev1alpha1.PostUpgradeVerification:       PostUpgradeVerification,
+			upgradev1alpha1.RemoveMaintWindow:             RemoveMaintWindow,
+			upgradev1alpha1.PostClusterHealthCheck:        PostClusterHealthCheck,
+		}
+	})
+
+	return upgrader
+}
+
+var postFixes = map[string]func(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error){
+	"4.3.0": func(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
```

Comment:
Also I don't think this should be committed as 4.3 is completed

Reply:
that's a good question, do you think we should remove that step for now? and we can revisit if we have a better solution in case of we do need post updates in future?